Add REVE model

jonathanlys01 opened the pull request:
This is a first attempt at integrating the REVE model to braindecode.
More documentation will follow once the implementation and usage are validated by the team.

To verify model instantiation and Hugging Face loading, you can run the following (a separate PR will add proper tests if this proves sufficient):

```python
from braindecode.models import REVE
import torch

n_chans = 64
n_times = 1000
batch_size = 8

eeg = torch.randn(batch_size, n_chans, n_times)
pos = torch.randn(batch_size, n_chans, 3)

# Test with random model
model = REVE()
out = model(eeg, pos)

# Load pretrained and use positions
model: REVE = REVE.from_pretrained("brain-bzh/reve-base", cache_dir="./cache")
channel_names = [f"E{i + 1}" for i in range(n_chans)]
pos = model.get_positions(channel_names).expand(eeg.size(0), -1, -1)  # (batch_size, channels, 3)
out = model(eeg, pos)
```

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,525 @@
+    FLASH_AVALIABLE = False
+    print(
+        "flash_attn not found, install it with `pip install flash_attn` if you want to use it"
+    )
```
I'm personally against putting `flash-attn` as a dependency, as it has proven to be quite difficult to install, and because its build is cuda version/torch version dependent. In some cases (I'm thinking about older GPUs like Volta or Tesla generation), it's not even possible to use flash attention.

The current setup works, in case the user doesn't have flash attention installed, the attention backend switches to (in order) flash attention (implemented by torch), efficient attention and MATH. A naive torch backend can also be triggered if people want to extract the attention matrix.

bruAristimunha replied:
My point here is that it's not the way to make dependencies optional.

jonathanlys01 replied:
Ended up deleting the flash attention component.

bruAristimunha commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,525 @@
+                If False, applies the final layer and returns the processed output. Default is False.
+        Returns:
+            Union[torch.Tensor, list[torch.Tensor]]: The output tensor(s) from the model. If `return_output` is True,
+                returns the transformer output; otherwise, returns the output after the final layer.
```
Documentation formation

bruAristimunha commented:
Almost all the comments for the copilot are validated and need to be addressed, except for the test that we already covered with the integration test, which follows the Timm style.

There are some more style details about layer names and how to become fully compatible with braindecode and still be able to load the weights, but I can address them when everything is resolved from the copilot initial review and the tests are green.

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,568 @@
+
+def patch_embedding(embed_dim, patch_size):
+    to_patch_embedding = nn.Sequential(nn.Linear(patch_size, embed_dim))
+    return to_patch_embedding
```
I tested this, it silently breaks the parameter loading of the patch embedding parameters.
I again highly support the introduction of tests that will check for the equality of the output with the hugging face model and the braindecode version.
 Something like:
```python
from braindecode.models import REVE
import torch
from transformers import AutoModel


batch_size = 2
n_chans = 32
n_times = 1000
n_outputs = 10
cache_dir = "./cache"

ch_list = [f"E{i + 1}" for i in range(n_chans)]
eeg_input = torch.randn(batch_size, n_chans, n_times)

model_braindecode: REVE = REVE.from_pretrained(
    "brain-bzh/reve-base",
    cache_dir=cache_dir,
    n_times=n_times,
    n_chans=n_chans,
    n_outputs=n_outputs,
)
pos_braindecode = model_braindecode.get_positions(ch_list)
pos_braindecode = pos_braindecode.unsqueeze(0).repeat(batch_size, 1, 1)

model_hf = AutoModel.from_pretrained(
    "brain-bzh/reve-base",
    cache_dir=cache_dir,
    trust_remote_code=True,
)
pos_bank_hf = AutoModel.from_pretrained(
    "brain-bzh/reve-positions",
    cache_dir=cache_dir,
    trust_remote_code=True,
)
pos_hf = pos_bank_hf(ch_list)
pos_hf = pos_hf.unsqueeze(0).repeat(batch_size, 1, 1)

# return_output is True to bypass the final layer
output_braindecode = model_braindecode(eeg_input, pos_braindecode, return_output=True)
output_hf = model_hf(eeg_input, pos_hf, return_output=True)

print("positions allclose:", torch.allclose(pos_braindecode, pos_hf, atol=1e-6))
print("outputs allclose:", torch.allclose(output_braindecode[-1], output_hf[-1], atol=1e-6))
```

jonathanlys01 commented:
I agree with all copilot comments about the style, there's just the small issue about the patch embedding layer being a Sequential with a single element, but replacing it with a plain linear silently breaks the loading of the parameters through hugging face (see [this comment](https://github.com/braindecode/braindecode/pull/866#discussion_r2609009597)).

jonathanlys01 commented:
Added the test, you will probably have to add a hugging face token in the GitHub secrets because the repository for REVE is gated.

bruAristimunha commented:
here is the braindecode org at hugging face: https://huggingface.co/braindecode

jonathanlys01 commented:
Hello @bruAristimunha , I've worked on the integration a bit:
- added the test for the model loading but the secret can't be loaded since it's an external PR, it works locally though.
- added REVE to the summary.csv, let me know if you agree on the labels (currently `Foundation Model` and `Attention/Transformer`.

Let me know if there is anything left to do before merging.

bruAristimunha commented:
Can you just double check the comments from copilot first revision? If you think it is okay, just resolved.

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,568 @@
+            size=self.patch_size,
+            step=self.patch_size - self.overlap_size,
+        )
+        batch_size, channel, heights, _n_patches = patches.shape
```
Done.

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,568 @@
+        super().__init__()
+        inner_dim = head_dim * heads
+        self.heads = heads
+        self.scale = head_dim**-0.5
```
Done

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,568 @@
+        Union[torch.Tensor, list[torch.Tensor]]
+            The output tensor(s) from the model. If `return_output` is True,
+            returns the transformer output; otherwise, returns the output after the final layer.
+        """
```
Added to the doc.

jonathanlys01 commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,568 @@
+            nn.Flatten(),
+            nn.LayerNorm(final_dim),
+            nn.Linear(final_dim, self.n_outputs),
+        )
```
Done

PierreGtch commented on braindecode/models/reve.py:
```diff
@@ -0,0 +1,618 @@
+
+        self.default_pos = None
+        if chs_info is not None:
+            self.default_pos = self.get_positions([ch["ch_name"] for ch in chs_info])
```
Why not directly using the positions provided in chs_info? Where did the positions come from in your channel bank?
Something like this:
```suggestion
            self.default_pos = self.get_positions([ch["loc"][:3] for ch in chs_info])
```

Or maybe a hybrid solution such that we also cover the case where the channel names are non-informative but positions are provided

jonathanlys01 replied:
The positions have been manually realigned to be consistent across montages. It's still possible to provide raw positions through the `pos` argument of the forward, but the recommended way to get the 3D coordinates is to use the `get_positions`, that takes a list of channel names as an input.

bruAristimunha commented:
I made one small change when I was trying to create the redirect link for the doc. Now the PR got moved for this: #876, but @jonathanlys01 still has the writing access. 

Sorry for my git issue.
